core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -57,6 +61,38 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
         return to.getLastTerrain().movementCost.toFloat() + extraCost // no road
     }
 
+    /** Returns whether the movement between the adjacent tiles [from] and [to] is affected by Zone of Control */
+    private fun isMovementAffectedByZoneOfControl(from: TileInfo, to: TileInfo, civInfo: CivilizationInfo): Boolean {
+        // Sources:
+        // - https://civilization.fandom.com/wiki/Zone_of_control_(Civ5)
+        // - https://forums.civfanatics.com/resources/understanding-the-zone-of-control-vanilla.25582/
+        //
+        // Enemy military units exert a Zone of Control over the tiles surrounding them. Moving from
+        // one tile in the ZoC of an enemy unit to another tile in the same unit's ZoC expends all
+        // movement points. Land units only exert a ZoC against land units. Sea units exert a ZoC
+        // against both land and sea units. Cities exert a ZoC as well, and it also affects both
+        // land and sea units. Embarked land units do not exert a ZoC. Finally, units that can move
+        // after attacking are not affected by zone of control if the movement is caused by killing
+        // a unit. This last case is handled in the movement-after-attacking code instead of here.
+        //
+        // We only need to check the two shared neighbors of [from] and [to]: the way of getting
+        // these two tiles can perhaps be optimized.
+        return from.neighbors.any{
+            to.neighbors.contains(it) && (
```

Comment:
This can indeed be optimised.

The neighbors.contains() is a sequence iteration, whereas the first check of both sides of the 'or' are equality comparisons, which are much faster, so even by just switching the order we should be able to get 'no' answers faster

Reply:
Switching the order is indeed probably faster! I have done so.

What I mostly meant with that comment, though, is that there is probably some way to define a `getSharedNeighbors` function that is more efficient than `from.neighbors.any{to.neighbors.contains(it)}`, probably by using some fancy hex math, but I didn't quite feel like diving into that.

Reply:
Also correct, and probably even vitally important... This function is what takes about 50% of the time on NextTurn().
So let's see:
There are 3 "variations": 
(0,0)<->(0,1):  (1,0), (-1,1)
(0,0)<->(1,0):  (0,1), (1,-1)
(0,0)<->(1,-1): (0,1), (-1,0)

So that's like:

fun adjacentTiles(tile1,tile2):
  if(tile1.x>tile2.x || tile1.x==tile2.x) return adjacentTiles(tile2,tile1) // switch order
  var diff = tile1.location.cpy().sub(tile2.location)
  val adjacentTile1 = tile1.cpy()
  val adjacentTile2 = tile1.cpy()
  if(diff.x==0 && diff.y==1){ adjacentTile1.x++; adjacentTile2.x--; adjacentTile2.y++}
  else if(diff.y==1 && diff.x==0) { adjacentTile1.y++; adjacentTile2.x++; adjacentTiley-- }
  else {adjacentTile1.y++; adjacentTile2.x--}
  return sequenceOf(adjacentTile1, adjacentTile2)

It's worth testing the actual performance difference - say by running each of them 1000 times and checking the timing of the runs.

Reply:
As determined by @SomeTroglodyte, the edge-aware version of this `adjacentTiles` implementation surprisingly performs worse than my initial naive approach.